File: bin/first_problem.rb
require 'json'

input_data = [
  {
    "name": "Alice",
    "email": "alice@gmail.com"
  },
  {
    "name": "Bob",
    "email": "bob@yahoo.com"
  },
  {
    "name": "Carol",
    "email": "carol@gmail.com"
  }
]

grouped_data = input_data.group_by { |data| data[:email].to_s.split('@')[1] }

mapped_data = grouped_data.each_with_object({}) { |(k, v), h| h[k] = v.map { |a| a[:name] } }

print JSON.pretty_generate({ data: mapped_data })

File: quizzes/quizzesV1/quizzes2/first_problem.js
const users = [
  { name: "Alice", email: "alice@gmail.com" },
  { name: "Bob", email: "bob@yahoo.com" },
  { name: "Carol", email: "carol@gmail.com" },
];

let groupedUsers = {};
for (let i = 0; i < users.length; i++) {
  const user = users[i];
  const domain = user?.email.split("@")[1];

  (groupedUsers[domain] ??= []).push(user?.name);
}

console.log(JSON.stringify({ groupedUsers }, null, 2));
